### ISSUE TITLE
Formatter indentation bug

### ISSUE THREAD

**Issue by Sija** [crystal-lang/crystal#5193]
Extracted from #5146.

```cr
[1, 2, 3]
  .each do |n|
    n
  end
```

Gets formatted to:

```cr
[1, 2, 3]
  .each do |n|
  n
end
```

/cc @MakeNowJust

**Comment by asterite** [comment]
Not a bug. The expression starts at indent 0 and the block is at indent 2. Look:

```crystal
# This is OK
[1, 2, 3].each do |n|
  n
end

# Now I just put a newline after `]`, still OK
[1, 2, 3]
  .each do |n|
  n
end
```

**Comment by straight-shoota** [comment]
Maybe it's not a bug but what about a feature request?

I think this example looks bad. It would make more sense to increase the indent of the next line after a `do` and close the block at the same indent as the line with `do` regardless of where the expression started.

**Comment by asterite** [comment]
Okay, then we have this:

```crystal
[1, 2, 3].map { |x| x - 1 }
         .select { |x| x > 1 }
         .each do |x|
  puts 1
end
```

Do we want it to format like this:

```crystal
[1, 2, 3].map { |x| x - 1 }
         .select { |x| x > 1 }
         .each do |x|
            puts 1
end
```

Or maybe it's already wrong that `.` is indent too far, so maybe it should be like this:

```crystal
[1, 2, 3].map { |x| x - 1 }
  .select { |x| x > 1 }
  .each do |x|
     puts 1
end
```

Or maybe like this:

```crystal
[1, 2, 3].map { |x| x - 1 }
         .select { |x| x > 1 }
         .each do |x|
            puts 1
         end
```

In the end, it boils down to user preference, something that the formatter doesn't allow.

But yes, we can continue discussing what's the best formatting here.

**Comment by Sija** [comment]
For my taste `end` should be on the same line as starting `do`, in the same vein as in other places.

**Comment by RX14** [comment]
@Sija uhh, you mean?

```cr
foo do |x|
      bar
    end
```

I think thats a terrible idea

**Comment by straight-shoota** [comment]
I think he means the `end` should have the same indent as *the line* including `do`. And the lines in between should be increased indent by one.

```crystal
[1, 2, 3]
  .each do |n|
    n
  end

foo do |x|
  bar
end

[1, 2, 3].map { |x| x - 1 }
         .select { |x| x > 1 }
         .each do |x|
            puts 1
         end
```

I'd say this is the most logical.

**Comment by Sija** [comment]
@straight-shoota yep, u get my drift.

**Comment by MakeNowJust** [comment]
All cases work fine!

![2017-10-28 0 03 12](https://user-images.githubusercontent.com/6679325/32110789-7017395e-bb73-11e7-8690-41d861003e33.png)

I'll create a new pull request.

**Comment by luislavena** [comment]
```crystal
[1, 2, 3].map { |x| x - 1 }
         .select { |x| x > 1 }
         .each do |x|
            puts 1
         end
```

The problem with the above is that a small change on position of the first call will cause a ripple effect on all the lines of the block:

```crystal
[1, 2, 3, 4, 5].map { |x| x - 1 }
               .select { |x| x > 1 }
               .each do |x|
                  puts 1
               end
```

Adding noise to the diffs. :confused:

**Comment by straight-shoota** [comment]
You can just ignore whitespace in the diff... And this style with high indents should certainly not be promoted.

You can write it better this way:
```crystal
numbers = [1, 2, 3, 4, 5]
numbers = .map { |x| x - 1 }
          .select { |x| x > 1 }
          .each do |x|
            puts 1
          end
```

**Comment by faultyserver** [comment]
A style that I've been using recently is to double-indent the lines before the block body starts, but keep everything else the same:

```crystal
[1, 2, 3, 4, 5]
    .map { |x| x - 1 }
    .select { |x| x > 1 }
    .each do |x|
  puts 1
end
```

I like this format because it keeps the body from being arbitrarily lengthened by leading whitespaced, and (imo) makes it more clear where the method chaining ends and the block body actually starts.

Particularly, I think this looks better for nested blocks:

```crystal
[1, 2, 3, 4, 5]
    .map { |x| x - 1 }
    .select { |x| x > 1 }
    .each do |x|
  puts 1

  [1, 2, 3, 4, 5]
      .map { |x| x - 1 }
      .select { |x| x > 1 }
      .each do |x|
    puts 1
  end
end
```

**Comment by jhass** [comment]
Let's call  this closed via #5234, everything else seems too opinionated.
